File: backend/models/order.py
"""
Order model for buy/sell water orders
"""

from sqlalchemy import Column, Integer, String, Float, DateTime, Enum, ForeignKey
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
import enum

from core.database import Base

class OrderType(str, enum.Enum):
    BUY = "buy"
    SELL = "sell"

class OrderStatus(str, enum.Enum):
    OPEN = "open"
    PARTIALLY_FILLED = "partially_filled"
    FILLED = "filled"
    CANCELLED = "cancelled"

class Order(Base):
    __tablename__ = "orders"
    
    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    
    # Order details
    order_type = Column(Enum(OrderType), nullable=False)
    quantity_af = Column(Float, nullable=False)  # Acre-feet
    filled_quantity_af = Column(Float, default=0.0)
    price_per_af = Column(Float, nullable=False)  # USD per acre-foot
    
    # Location/basin for matching
    basin = Column(String, nullable=False)
    
    # Status
    status = Column(Enum(OrderStatus), default=OrderStatus.OPEN)
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    filled_at = Column(DateTime(timezone=True))
    
    @property
    def remaining_quantity(self) -> float:
        return self.quantity_af - self.filled_quantity_af

File: ios/WaterXchange/Models/Order.swift
//
//  Order.swift
//  WaterXchange
//

import Foundation

enum OrderType: String, Codable, CaseIterable {
    case buy = "buy"
    case sell = "sell"
    
    var displayName: String {
        switch self {
        case .buy: return "Buy"
        case .sell: return "Sell"
        }
    }
    
    var color: String {
        switch self {
        case .buy: return "BuyGreen"
        case .sell: return "SellRed"
        }
    }
}

enum OrderStatus: String, Codable {
    case open = "open"
    case partiallyFilled = "partially_filled"
    case filled = "filled"
    case cancelled = "cancelled"
    
    var displayName: String {
        switch self {
        case .open: return "Open"
        case .partiallyFilled: return "Partial"
        case .filled: return "Filled"
        case .cancelled: return "Cancelled"
        }
    }
    
    var color: String {
        switch self {
        case .open: return "StatusOpen"
        case .partiallyFilled: return "StatusPartial"
        case .filled: return "StatusFilled"
        case .cancelled: return "StatusCancelled"
        }
    }
}

struct Order: Codable, Identifiable {
    let id: Int
    let orderType: OrderType
    let quantityAF: Double
    let filledQuantityAF: Double
    let pricePerAF: Double
    let basin: String
    let status: OrderStatus
    let createdAt: Date
    
    enum CodingKeys: String, CodingKey {
        case id, basin, status
        case orderType = "order_type"
        case quantityAF = "quantity_af"
        case filledQuantityAF = "filled_quantity_af"
        case pricePerAF = "price_per_af"
        case createdAt = "created_at"
    }
    
    var remainingQuantity: Double {
        quantityAF - filledQuantityAF
    }
    
    var totalValue: Double {
        quantityAF * pricePerAF
    }
}

struct CreateOrderRequest: Codable {
    let orderType: String
    let quantityAF: Double
    let pricePerAF: Double
    
    enum CodingKeys: String, CodingKey {
        case orderType = "order_type"
        case quantityAF = "quantity_af"
        case pricePerAF = "price_per_af"
    }
}

struct OrderListResponse: Codable {
    let orders: [Order]
    let total: Int
}
